### PR TITLE
<StyledNestableList/> - new component

### PR DESCRIPTION
**Pull request by AndriiTsarenko**
### 🔦 Summary

...

### ✅ Checklist

- [ ] 👨‍💻 API change is approved by the UI Infra Developers 
- [ ] 👨‍🎨 UX change is approved by the Design System UX 
- 📚 Change is documented
  - [ ] Story
  - [ ] API description
  - [ ] Cheatsheet
  - [ ] Other (explain)
- 🔬 Change is tested
  - [ ] Component
  - [ ] Visual test

### CODE REVIEW

**Comment by WixBuildServer** [comment]
### 🤖 Hi, I'm #Janet!

  A deploy preview of your PR was successfully released 🖼.

  You can find it under one of the following URLs:
  * https://preview.wix-style-react.com/pr_6231

File: src/StyledNestableList/docs/index.story.js [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+    }),
+
+    tabs([
+      tab({
+        title: 'Description',
+        sections: [
+          description({
+            title: 'Description',
+            text:
+              'This line here should briefly describe component in just a sentence or two. It should be short and easy to read.',
```
Comment (EgoziE) [inline src/StyledNestableList/docs/index.story.js:66]:
```suggestion
              'A styled list with drag and drop and nesting capabilities.',
```

File: stories/index.js [hunk 2]
```diff
@@ -216,3 +216,6 @@ require('../src/Themes/editorX/docs');
 
 // BusinessDashboard
 require('../src/Themes/businessDashboard/docs');
+
+// TODO: move to correct position
+require('../src/StyledNestableList/docs/index.story');
```
Comment (EgoziE) [inline stories/index.js:221]: Move it up under Components API

File: src/StyledNestableList/StyledNestableList.uni.driver.d.ts [hunk 3]
```diff
@@ -0,0 +1,7 @@
+import { BaseUniDriver } from 'wix-ui-test-utils/unidriver';
+
+export interface StyledNestableListUniDriver extends BaseUniDriver {
+  getCountText(): Promise<string>;
+  clickButtonTimes(times: number): Promise<void>;
+  getButtonText(): Promise<string>;
```
Comment (EgoziE) [inline src/StyledNestableList/StyledNestableList.uni.driver.d.ts:6]: Should be removed

File: src/StyledNestableList/StyledNestableList.uni.driver.js [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+        await buttonElement.click();
+      }
+    },
+
+    /**
+     * Gets the button text
+     * @returns {Promise<string>}
+     */
+    getButtonText: () =>
+      findByHook(base, dataHooks.styledNestableListButton).text(),
```
Comment (EgoziE) [inline src/StyledNestableList/StyledNestableList.uni.driver.js:35]: Should be removed

File: src/StyledNestableList/StyledNestableList.st.css [hunk 5]
```diff
@@ -0,0 +1,83 @@
+:import {
+  -st-from: '../Foundation/stylable/colors.st.css';
+  -st-named: D60;
+}
+
+.root {
+  background-color: white;
```
Comment (EgoziE) [inline src/StyledNestableList/StyledNestableList.st.css:7]: Please change to D80

File: src/StyledNestableList/StyledNestableList.st.css [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+}
+
+.actionItem {
+  margin-left: 84px;
+}
+
+.lastItem {
+}
+.haveChildren {
+}
```
Comment (EgoziE) [inline src/StyledNestableList/StyledNestableList.st.css:23]: Empty classes are removed on build, please remove them or add any style property to them

File: src/StyledNestableList/StyledNestableList.js [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+import Box from '../Box';
+
+import { Arrow } from './Arrow';
+import { TextButton } from '../index';
+import Add from 'wix-ui-icons-common/Add';
+import { isFistItem, isLastItem, isRootItem } from '../NestableList/helpers';
+
+/** TBD */
+class StyledNestableList extends React.PureComponent {
+  renderArrow = ({ isLastChild = false, isPreview = false, depth } = {}) => {
```
Comment (EgoziE) [inline src/StyledNestableList/StyledNestableList.js:18]: depth is not used, should it be?

File: src/StyledNestableList/StyledNestableList.js [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+  renderItem = options => {
+    const { isPlaceholder, depth, isPreview, siblings, item } = options;
+    const { id, children, isCollapsed, ...rest } = item;
+    const isLastChild = isLastItem(siblings, item);
+    return (
+      <Box>
+        {!isRootItem(depth) &&
+          this.renderArrow({ isLastChild, isPreview, depth })}
+        <Box
+          display={'block'}
```
Comment (EgoziE) [inline src/StyledNestableList/StyledNestableList.js:87]: Change to a simple div with a className

File: src/StyledNestableList/StyledNestableList.js (hunk 7 above)
Comment (EgoziE) [inline src/StyledNestableList/StyledNestableList.js:18]: Add an underscore to all functions to mark them as private
e.g.
renderArrow => _ renderArrow

File: src/StyledNestableList/StyledNestableList.js [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+  /** triggers in case when add new item button is clicked */
+  onAddItem: PropTypes.func,
+  /** triggers when order or nesting is changed */
+  onChange: PropTypes.func,
+  /** Maximum depth of the node three */
+  maxDepth: PropTypes.number,
+  /** Allow drop node only on it's own depth */
+  preventChangeDepth: PropTypes.bool,
+  /** Nodes to render */
+  items: PropTypes.arrayOf(NodeShape),
```
Comment (EgoziE) [inline src/StyledNestableList/StyledNestableList.js:168]: Unfortunately our AutoDocs doesn't know how to display complex types, so just write it down as a comment
```suggestion
  /**
   * * __id__ - Specifies the item id
   * * __isCollapsed__ - Whether to render the items children
   * * __children__ - A list of children of type item
   * * All other TableListItem props
   */
  items: PropTypes.arrayOf(NodeShape),
```

**Review by EgoziE — CHANGES_REQUESTED** [review]
Thanks for this PR! You did an amazing job!

I wrote some inline comments, just one more thing which is not critical but if you have the time it can be nice to have:
We use Stylable as our css preprocessor, so some best practices you know from scss doesn't apply here.
So instead of using "classnames" for conditional styling, we use Stylable states.
For example:
Instead of:
```
        <div
          className={classNames(classes.verticalArrow, {
            [classes.lastItemVerticalArrow]: isLastChild,
            [classes.previewArrows]: isPreview,
          })}
        >
```

You should write:
```
        <div
          className={st(classes.verticalArrow, {
            lastChild: isLastChild,
            preview: isPreview,
          })}
        >
```
And the css should be something like:
```
.verticalArrow {
  -st-states: lastChild, preview;
}

.verticalArrow:lastChild { ... }

.verticalArrow:preview { ... }
```

**Comment by milkyfruit** [comment]
That's amazing, Andrii!
I've a bunch of questions, please don't be scared.
- [ ] Component has a prop `preventChangeDepth`, however it still allows to leave the parent and move to another parent. Do you think we should have another prop, which would restrict that?
- [ ] `addItemLabel` Does it allow to apply different labels to separate add items? I see there might be a need for that.
- [ ] If nesting is restricted (maxDepth is set to 1), how can I expose "add Item"?

**Review by justinara — CHANGES_REQUESTED** [review]
Great job so far! But I have quite a few comments from my side.. some are minor and small adjustments, some a little more complex:

- [x] Picked up item should have a hover style (table list item with B50 background colour and drag handle in colour B10)
- [x] Picked up item is missing drop shadow https://drive.google.com/open?id=1FrKXSjJGD3xZRh4gwcnWBJzlWQiXYL1z
- [x] Picked up item has a 1px border from all sides. Should not have that
Screenshot at Nov 16 15-33-48

- [ ] Cursor should be `grabbing` type  once item is picked up and dragged

image

- [x] At the moment I'm not able to hide bottom divider line for last item on the list using table list item prop `showDivider={false}` - it doesn't work :/

- [x] Children items should collapse once parent item is picked up: https://drive.google.com/open?id=19QxLIXD2cOKK9GFwxAJUrUCyrFp1yoEg - this should include "Add new" as well

- [x] When changing the level of an item, arrow icon should be displayed on a grey drop target background: 

image. animated preview: https://drive.google.com/open?id=1yYIDS1jkKLT07S4eG6-MJZr146Y5qIMO

- [ ] Missing animations in general. Items should slide vertically as per preview here: https://drive.google.com/open?id=1GYnjdaKI6sMkLXpteDOl01pfFE5ZYXZa. Sliding on y axis starts once picked up item reaches the middle of an item below/above it.

- [ ] `Prevent nodes from dragging with drag disabled icon` example doesn't work as expected. If item has `draggable`: true and `dragDisabled`:true,  you should not be able to drag and change the location of that specific item. But you can at the moment: 

image

- [ ] Drop target change to third level happens too early, it it displayed once second level should be shown only: 

image 
Please follow these guidelines : 
image

Please also check animations section in zeplin specs, and previews here: https://drive.google.com/open?id=1x54N2459v-6aCZbTXfwXLhMJGRFh0RAA

**Comment by justinara** [comment]
Also, another question - would it be possible to add keyboard accessibility in this effort as well? Drag grip can be focused once user navigates through page with TAB key click:
image

How can it be picked up and reordered?
1. Item is picked up by ENTER key click (keeps focus on drag grip, has a hover style of list item and a shadow): 
image
2. You can move it up on the same level with ARROW UP
3. You can move it down on the same level with ARROW DOWN
4. You can make it a child item by pressing on ARROW RIGHT
5. You can change its level from child to parent by pressing ARROW LEFT
6. Item is released / submitted to new spot on ENTER click as well.

Thoughts on how much development time that would require?

**Comment by AndriiTsarenko** [comment]
Draggable, drag disabled, and read-only I did like so:
On item:
draggable: true - draggable and with drag icon
draggable: false - is not draggable and disabled drag icon
by default draggable is true.

On component
readOnly - readOnly mode
@milkyfruit

**Comment by milkyfruit** [comment]
> Draggable, drag disabled, and read-only I did like so:
> On item:
> draggable: true - draggable and with drag icon
> draggable: false - is not draggable and disabled drag icon
> by default draggable is true.
> 
> On component
> readOnly - readOnly mode
> @milkyfruit

Looks great, @AndriiTsarenko ! We can move the story to WIP folder and then merge it:
image
This way you'll be able to start using in your project, and once @justinara comments are resolved the story will be moved to Components folder.

File: src/StyledNestableList/StyledNestableList.st.css (hunk 6 above)
Comment (AndriiTsarenko) [reply src/StyledNestableList/StyledNestableList.st.css:31]: Done in 
https://github.com/wix/wix-style-react/pull/6416

**Pull request merged**